### PR TITLE
[FLINK-20837] Refactor dynamic SlotID

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

Refactor dynamic SlotID.
In FLINK-14189, we specified that the SlotID whose slot number is -1 represents the dynamic slot. However, the SlotID with a negative slot number will be treated as invalid in other components, e.g. SlotPool and SlotManager. We have to remove a lot of sanity checks with this change, which is error-prone. The evidence is that there are still some sanity checks that have not been removed, e.g. sanity checks in AllocatedSlotInfo.

In this PR, we refactor the dynamic SlotID:
- In SlotManager, it still uses -1 as slot number when allocating dynamic slot.
- In TM, when allocating dynamic slots, it will asssign an increasing number larger than the numberSlots as the slot number.

## Verifying this change

This change is already covered by existing tests:
- TaskExecutorTest
- TaskSlotTableImplTest

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java [hunk 1]
```diff
@@ -329,7 +339,7 @@ public boolean allocateSlot(
                         jobId,
                         allocationId,
                         memoryVerificationExecutor);
-        if (index >= 0) {
+        if (index < numberSlots) {
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java:342]: Now since the dynamic slots also have unique indexes, we can also insert them into `taskSlots`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java [hunk 2]
```diff
@@ -463,7 +473,7 @@ public boolean isAllocated(int index, JobID jobId, AllocationID allocationId) {
         TaskSlot<T> taskSlot = taskSlots.get(index);
         if (taskSlot != null) {
             return taskSlot.isAllocated(jobId, allocationId);
-        } else if (index < 0) {
+        } else if (index >= numberSlots) {
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java:476]: If we also insert dynamic slot to the `taskSlot`, we won't need this `else-if` branch anymore.

**Review by xintongsong — CHANGES_REQUESTED** [review]
Thanks for the PR, @KarmaGYZ.
I've left a few comments. Please take a look.

**Comment by xintongsong** [comment]
And please take a look at the compiling error reported by AZP.

**Comment by KarmaGYZ** [comment]
Thanks for the review @xintongsong . PR updated.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java [hunk 3]
```diff
@@ -288,22 +288,13 @@ public boolean allocateSlot(
...
         TaskSlot<T> taskSlot = allocatedSlots.get(allocationId);
         if (taskSlot != null) {
-            LOG.info("Allocation ID {} is already allocated in {}.", allocationId, taskSlot);
-            return false;
-        }
-
-        if (taskSlots.containsKey(index)) {
-            TaskSlot<T> duplicatedTaskSlot = taskSlots.get(index);
+            return isDuplicatedSlot(taskSlot, jobId, resourceProfile, index);
+        } else if (!isIndexAlreadyTaken(index)) {
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java]: `!` should be removed.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java [hunk 4]
```diff
@@ -288,22 +288,13 @@ public boolean allocateSlot(
...
             LOG.info(
-                    "Slot with index {} already exist, with resource profile {}, job id {} and allocation id {}.",
+                    "The static slot with index {} is already assigned to another allocation with id {}.",
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java:301]: Not sure about exposing the concept *static* slot here.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java [hunk 5]
```diff
@@ -277,27 +280,31 @@ public boolean allocateSlot(
...
+        } else if (isIndexAlreadyTaken(index)) {
             LOG.info(
                     "The static slot with index {} is already assigned to another allocation with id {}.",
                     index,
                     taskSlots.get(index).getAllocationId());
             return false;
         }
 
-        resourceProfile = index >= 0 ? defaultSlotResourceProfile : resourceProfile;
+        resourceProfile = index < numberSlots ? defaultSlotResourceProfile : resourceProfile;
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java:307]: There are 3 occurrences of `index < numberSlots` and 1 occurrence of `index >= numberSlots` in this file.
Let's deduplicate it with a util method.

**Review by xintongsong — COMMENTED** [review]
It might be better to also add test cases for the duplicate/inconsistent slot allocation, for both static and dynamic slots. (i.e., the logics that the first commit touches.)

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImplTest.java [hunk 6]
```diff
@@ -115,11 +115,11 @@ public void testRetrievingAllActiveSlots() throws Exception {
...
 
     /**
-     * Tests that redundant slot allocation with the same AllocationID to a different slot is
-     * rejected.
+     * Tests that inconsistent static slot allocation with the same AllocationID to a different slot
+     * is rejected.
      */
     @Test
-    public void testRedundantSlotAllocation() throws Exception {
+    public void testInconsistentStaticSlotAllocation() throws Exception {
```
Comment (xintongsong) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImplTest.java:122]: This covers the case trying to allocate the same allocation id to different slot index.
I think we need to also cover the case that different allocation ids trying to take the same slot index

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImplTest.java [hunk 7]
```diff
@@ -137,6 +137,68 @@ public void testRedundantSlotAllocation() throws Exception {
...
+
+            Iterator<TaskSlot<TaskSlotPayload>> allocatedSlots =
+                    taskSlotTable.getAllocatedSlots(jobId);
+            assertThat(allocatedSlots.next().getIndex(), is(0));
+            assertThat(allocatedSlots.hasNext(), is(false));
+        }
+    }
+
+    @Test
+    public void testDuplicateDynamicSlotAllocation() throws Exception {
```
Comment (xintongsong) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImplTest.java:185]: I think we should also verify that the second call of `allocateSlot` does not replace the slot allocated by the first call.

**Review by xintongsong — COMMENTED** [review]
Thanks for updating the PR, @KarmaGYZ.
The changes already look quite good to me. I have only a few minor comments regarding the test verifications.

**Comment by KarmaGYZ** [comment]
PR updated.

**Review by xintongsong — APPROVED** [review]
Thanks @KarmaGYZ. LGTM.
Merging once AZP gives green light.
